### ISSUE TITLE
Adding Audio on http page , and adding record activity like mp3 format ?

### ISSUE THREAD

**Issue by khanfar** [boatbod/op25#37]
Hi , is there way to push audio in http web view ? If no hope to add this in future and is there easy way to record voice activity like mp3 or something else in one file . ?

**Comment by s3m1s0n1c** [comment]
You could put a Audio Player on the page that connects to a icecast stream locally or remote..

**Comment by hayden-t** [comment]
Both would be possible.

**Comment by k4jk** [comment]
I am working adding audio over websocket in a branch:

https://github.com/k4jk/op25/tree/http_audio

If anyone wants to try it out/test just add a web socket destination to your channel config like this:

"destination": "ws://0.0.0.0:9000",

The logic in main.js works like this:

If the host in destination is 0.0.0.0 or 127.0.0.1, the browser substitutes window.location.hostname automatically- so it will just try to load audio from the same address in the URL. 

If the host is an explicit IP like 192.168.1.100, the browser will load audio from the specified address as-is, in case you have a reason to do it that way

Click on the Headphones Icon between the tuning and channel buttons to start the audio to your browser.

**Comment by boatbod** [comment]
Awesome!  I shall take a look :)

**Comment by k4jk** [comment]
> Awesome! I shall take a look :)

Cool thanks! It now seems pretty stable after those changes to the C++ audio files but needs wider testing. I have tested with Firefox/Brave/Safari/Edge and seems ok so far.

**Comment by boatbod** [comment]
I've got things up and running and so far it's working very well with no hiccups or glitches :)

Presently it looks like you can only stream the current channel being viewed in the web terminal.  If possible, it'd be great if you could stream all channels simultaneously unless they have been explicitly paused by clicking on the headphone icon.  My test system monitor three counties individually (i.e. 3 channels) and all can be sending audio at the same time.  On rare occasions it can get too hectic to understand, but mostly it works out ok.

**Comment by k4jk** [comment]
OK good to hear, please keep fiddling with it. I noticed last night that is was flaky if you ctrl-c and quickly re-launch, which led to the latest fix in op25_audio.cc.

I agree on the simultaneous multi-channel audio, I tried it yesterday with a multi-system config and immediately noticed the limitations. I think it's pretty straightforward to change.

I was also thinking about an easy way to record through the UI, but I think it would be better to use Opus and MediaRecorder rather than try to do something with MP3 output. WebM/Opus would be really lightweight and have native support in every browser (except Safari I guess)

**Comment by boatbod** [comment]
Couple thoughts:
i. Once you have the multi-channel working, please go ahead and submit a pull request against my 'dev' branch so that I can pull your changes in and exercise them more completely.

ii. I suspect we might need to give some though to supporting WSS as well as WS.  I think this is going to be more of a server-side change, but it's my understanding that in order to establish a websocket connection to a non-local endpoint (i.e. something other than 127.0.0.1) you have to have a secure connection.

I'm not familiar with WebM/Opus, that said, recording probably ought to be a server function.

**Comment by k4jk** [comment]
Will do on the PR, should hopefully have it ironed out in the next day or two.

WS as implemented works across my Lan- I'm currently streaming from a headless mini pc in my shack to a laptop via wifi.

**Comment by k4jk** [comment]
Implemented in PR #278

**Comment by boatbod** [comment]
Is there a way to make it connect automatically when the page opens?  Ideally there would be a checkbox in the settings screen to enable/disable auto connect.

**Comment by k4jk** [comment]
Yeah a toggle should be able to be added to local storage I think. I'll look at adding that

**Comment by k4jk** [comment]
Added in  #279

**Comment by boatbod** [comment]
Everything has been merged in to dev, master and gr310 branches.

**Comment by boatbod** [comment]
Notes from the code merge:
**Auto-play at startup: Audio channels now start unmuted by default.**
    On the first user interaction with the page, the AudioContext is created
    and all connected channels begin playing automatically — no need to manually
    click the headphones icon for each channel. 
**_Some interaction with the page is required, due to browser autoplay security policy. A simple click anywhere on the UI suffices._**

**"Mute Browser Audio at Startup" setting: A new checkbox in the UI**
    Settings panel allows users to set the audio to be muted at startup, until specifically enabled per channel. The setting is saved to localStorage and persists across page refreshes.

**Comment by wgbecks** [comment]
I am finding that “Some interaction with the page is required” is almost always a requirement to actually receiving audio from the speaker. Simply

clicking on the UI screen doesn’t work, I have to click on the settings or channel up/down to unmute despite the icon showing active unmuted audio.

 

Additionally, the “Mute Browser Audio at Startup” does not remain persistent upon a reconnection/restart of the browser.

**Comment by k4jk** [comment]
Are you using Safari? Safari security for autoplay is more strict than other browsers, it won't allow AudioContext to be created unless an element is actually interacted with (Menu, Button, Input Box, etc.) I probably should have been more specific about that in the PR description. 

For the second problem, also smells like Safari, possibly strict settings for removing history and Local Storage when the browser is closed. Safari also has adaptive browser security settings that can change how data is stored on sites that are not visited often. 



> I am finding that “Some interaction with the page is required” is almost always a requirement to actually receiving audio from the speaker. Simply
> 
> clicking on the UI screen doesn’t work, I have to click on the settings or channel up/down to unmute despite the icon showing active unmuted audio.
> 
> 
> 
> Additionally, the “Mute Browser Audio at Startup” does not remain persistent upon a reconnection/restart of the browser.
> 
> 
>

**Comment by wgbecks** [comment]
No, not using Safari.  All testing so far has been on two different Windows-11 machine with Edge and Chrome browsers so far.

**Comment by k4jk** [comment]
OK thanks, I can't reproduce this on Edge/Windows 11. It sounds like some security policy is keeping Audio Context from being created without a interaction, or your browser is in a suspended state. Can you please do the following:

Open the page
Click in the UI somewhere 
Open developer tools (F12)
Click the Console tab in Developer Tools
In the console, (you should see a  ">"  prompt),  type this and hit Return:

console.log(typeof audioCtx !== 'undefined' && audioCtx !== null ? audioCtx.state : 'AudioContext not yet created');

Please let me know what it says. It should help us understand what is going on with AudioContext

If you want to paste the above command you might have to type "allow paste" first



> No, not using Safari.  All testing so far has been on two different Windows-11 machine with Edge and Chrome browsers so far.
> 
> 
>

**Comment by wgbecks** [comment]
James,

 

Below are the console results from tests using Microsoft Edge and Firefox browsers on Windows-11 24H2.

Please note that I get identical results when testing from the Firefox browser on Ubuntu 24.04 and Linux Mint 22.3.

 

Is there a way to code the WS such that browser interaction is not required?

 

Thanks!

 

Bill

 

 

Edge Version 147.0.3912.98 with no browser interaction: No speaker audio

main.js:2508 WebSocket audio connected for channel 0 : ws://192.168.0.50:9000/

console.log(typeof audioCtx !== 'undefined' && audioCtx !== null ? audioCtx.state : 'AudioContext not yet created');

VM65:1 AudioContext not yet created

undefined

 

Edge Version 147.0.3912.98 with tap on browser interaction: Have speaker audio

console.log(typeof audioCtx !== 'undefined' && audioCtx !== null ? audioCtx.state : 'AudioContext not yet created');

VM71:1 running

undefined

 

Firefox Version 150.0.1 with no browser interaction: No speaker audio

WebSocket audio connected for channel 0 : ws://192.168.0.50:9000/ main.js:2508:13

console.log(typeof audioCtx !== 'undefined' && audioCtx !== null ? audioCtx.state : 'AudioContext not yet created');

AudioContext not yet created debugger eval code:1:9

undefined

 

Firefox Version 150.0.1 with tap on browser interaction: Have speaker audio

console.log(typeof audioCtx !== 'undefined' && audioCtx !== null ? audioCtx.state : 'AudioContext not yet created');

running debugger eval code:1:9

undefined

**Comment by k4jk** [comment]
Thanks for testing. Just for clarity's sake, when you say this: 

"tap on browser interaction"

Do you mean you are just clicking somewhere that is NOT a menu or button, or are you actually clicking on a button/menu etc?


Thanks,

JK

> James,
> 
> 
> 
> Below are the console results from tests using Microsoft Edge and Firefox browsers on Windows-11 24H2.
> 
> Please note that I get identical results when testing from the Firefox browser on Ubuntu 24.04 and Linux Mint 22.3.
> 
> 
> 
> Is there a way to code the WS such that browser interaction is not required?
> 
> 
> 
> Thanks!
> 
> 
> 
> Bill
> 
> 
> 
> 
> 
> Edge Version 147.0.3912.98 with no browser interaction: No speaker audio
> 
> main.js:2508 WebSocket audio connected for channel 0 : ws://192.168.0.50:9000/
> 
> console.log(typeof audioCtx !== 'undefined' && audioCtx !== null ? audioCtx.state : 'AudioContext not yet created');
> 
> VM65:1 AudioContext not yet created
> 
> undefined
> 
> 
> 
> Edge Version 147.0.3912.98 with tap on browser interaction: Have speaker audio
> 
> console.log(typeof audioCtx !== 'undefined' && audioCtx !== null ? audioCtx.state : 'AudioContext not yet created');
> 
> VM71:1 running
> 
> undefined
> 
> 
> 
> Firefox Version 150.0.1 with no browser interaction: No speaker audio
> 
> WebSocket audio connected for channel 0 : ws://192.168.0.50:9000/ main.js:2508:13
> 
> console.log(typeof audioCtx !== 'undefined' && audioCtx !== null ? audioCtx.state : 'AudioContext not yet created');
> 
> AudioContext not yet created debugger eval code:1:9
> 
> undefined
> 
> 
> 
> Firefox Version 150.0.1 with tap on browser interaction: Have speaker audio
> 
> console.log(typeof audioCtx !== 'undefined' && audioCtx !== null ? audioCtx.state : 'AudioContext not yet created');
> 
> running debugger eval code:1:9
> 
> undefined
> 
> 
>

**Comment by wgbecks** [comment]
Correct.  I just one click on the UI (browser) frame.  Although any menu or button press works too.

**Comment by k4jk** [comment]
OK thanks for the clarity- if you are just clicking on any area in the UI to get audio to unmute that's essentially how it is supposed to work.

I am not really a Front End developer, but my understanding is that every modern browser has baked-in security to require a user to have some kind of contextual interaction with what is rendered before audio will be played. There's not really a way to force it to play in the javascript that is loaded, nor SHOULD that be done IMO. 

There may be some setting that could be changed in your browser settings to ignore the context requirements and autoplay audio anyway, I am not sure.



> Correct.  I just one click on the UI (browser) frame.  Although any menu or button press works too.
> 
> 
> 
> 
> 
> From: James Kirkham ***@***.***>
> Sent: Wednesday, May 6, 2026 1:10 PM
> To: boatbod/op25 ***@***.***>
> Cc: wgbecks ***@***.***>; Comment ***@***.***>
> Subject: Re: [boatbod/op25] Adding Audio on http page , and adding record activity like mp3 format ? ([#37](https://github.com/boatbod/op25/issues/37))
> 
> 
> 
>   <https://avatars.githubusercontent.com/u/5454677?s=20&v=4> k4jk left a comment ([boatbod/op25#37](https://github.com/boatbod/op25/issues/37)) <[#37 (comment)](https://github.com/boatbod/op25/issues/37#issuecomment-4390728483)>
> 
> Thanks for testing. Just for clarity's sake, when you say this:
> 
> "tap on browser interaction"
> 
> Do you mean you are just clicking somewhere that is NOT a menu or button, or are you actually clicking on a button/menu etc?
> 
> Thanks,
> 
> JK
>

**Comment by wgbecks** [comment]
Thanks, James.   

 

I’ve been searching and testing a number of methods to allow the audio context (autoplay) but so far, none of them work.  Likely old AI data

from previous builds of Windows/Edge.  I don’t know if it would help, but perhaps it would satisfy security to move op25’s UI off of the python

waitress and adapt to work with Apache or Nginx in support of https, certificates, and etc.

 

Bill

**Comment by k4jk** [comment]
The AudioContext security is contained completely on the client side in the application layer, it would still apply to whatever tech stack the HTTP server is running on, or whether or not TLS is enabled.

The python webserver could theoretically support SSL, I think the only change there would be to wrap the socket in a SSL context.

It looks to me like the big challenge with SSL would be the websocketpp class in the C++ code (op25_audio.cc). Regardless, enabling SSL would still not create the desired behavior with autoplay.

**Comment by boatbod** [comment]
The websocketpp code does support SSL but, but getting it running would require that op25 obtained a certificate during the install process, and this certificate would have to be specific for the domain it was operating under.  Certs are free through LetsEncrypt, but I fear the added install complexity would preclude many users from bothering with it.  Also, unless there is a tangible benefit of having SSL enabled, I'm not convinced the extra coding effort would be worth it.  SSL would also need to be turned on for the web page itself.

**Comment by k4jk** [comment]
Agree with all that @boatbod, I am not sure the juice is worth the squeeze. 

The install script could also possibly walk a user through a local self-signed cert generation, (bypassing the complexity of obtaining a cert from a real CA) but when probably 99.9% of users are just streaming across their LAN it's kind of pointless.

If someone really needs to stream data from op25 securely across the internet they can always use a VPN or something.

**Comment by k4jk** [comment]
We have a few conventional P25 systems around here, and I noticed that no info was being pulled into the webUI about them when configured to run in multi_rx.py. A websocket for them was connecting and audio was playing, but there was no mute control available etc. because the channel data never make into the client.

tk_p25.py was modified to process data from conventional frames, main.js was modified to handle the conventional data and display appropriately. 

PR #282

**Comment by wgbecks** [comment]
I’d definitely like to see support in the UI for conventional P25 systems.

**Comment by boatbod** [comment]
I've merged the Conventional pull request, but have been playing around a little bit with having multi_rx build a default trunking config if none exists.  I don't have a good way to do much testing on it, other than some replays of conventional traffic, so perhaps you both might give it a shot.

Key question is whether we need more basic data in the trunking system definition ("chans" entry).  In the following P25 Conventional config more stuff show in the (curses) terminal when I use this config rather than the raw module-only config.

Thoughts?
Feel free to tweak multi_rx.py line 587 to specify a better default config.

**Comment by k4jk** [comment]
Your latest change to multi_rx.py works as long as any reference to trunking is absent in a pure conventional config, OR you have a trunking node that calls the tk_p25.py module.


1) If you have a "trunking" node in the json with an empty channel array but the module set to "tk_p25.py" it works
2) If you have a "trunking" node in the json that is otherwise completely empty,  it breaks. (This is how the existing p25_conventional_example.json is configure, so if we want this we should probably change the example)

So in other words, if the trunking node was completely removed in a conventional-only config it seems to work fine

**Comment by boatbod** [comment]
I'll explore some more, but probably not until tonight.
G

ETA:- fixed that issue (presently in dev only)

**Comment by wgbecks** [comment]
Gents,

 

I have just testing Conventional P25 with this commit and it works perfectly.   See attached p25.json used for monitoring my ZUMspot (Ham Hotspot)

along with a copy of the logfile. 

 

Conventional Analog (FM) does not produce a websocket.  There appears to be a conflict with socket ports in that the underlying digital demodulator appears to get assigned ws 9000 that triggers the conflict for the analog demodulator.  Attached are my noaa_ws.json and noaa_ws.log files for review.

 

Thank you,

 

Bill, WA8WG

**Comment by k4jk** [comment]
@boatbod 

I had to make a one-line change to main.js to guard against a null trunking key when there is a conventional-only config #283 

> I'll explore some more, but probably not until tonight.
> G
> 
> ETA:- fixed that issue (presently in dev only)

**Comment by k4jk** [comment]
Hi Bill,

Can you open a new issue on github for the analog audio websocket issue? I think you are right, there is a conflict with the websocket so the fix will probably not be very trivial.

Also, it looks like your attachments didn't go through. You might be able to attach them to the github issue

**Comment by boatbod** [comment]
Browser audio now operational.
Please submit any additional bug reports as a new issue.
